protocol/x/subaccounts/keeper/isolated_subaccount.go

```diff
@@ -0,0 +1,144 @@
+// The input subaccounts must be settled.
+//
+// Returns a `success` value of `true` if all updates are valid.
+// Returns a `successPerUpdates` value, which is a slice of `UpdateResult`.
+// These map to the updates and are used to indicate which of the updates
+// caused a failure, if any.
+func (k Keeper) checkIsolatedSubaccountConstraints(
+	ctx sdk.Context,
+	settledUpdates []settledUpdate,
+	perpetuals []perptypes.Perpetual,
+) (
+	success bool,
+	successPerUpdate []types.UpdateResult,
+	err error,
+) {
+	success = true
+	successPerUpdate = make([]types.UpdateResult, len(settledUpdates))
+	var idOfSettledUpdates = make(map[types.SubaccountId]struct{})
+	var perpIdToMarketType = make(map[uint32]perptypes.PerpetualMarketType)
+
+	for _, perpetual := range perpetuals {
+		perpIdToMarketType[perpetual.GetId()] = perpetual.Params.MarketType
+	}
+
+	for i, u := range settledUpdates {
+		_, exists := idOfSettledUpdates[*u.SettledSubaccount.Id]
+
+		if exists {
+			return false, nil, types.ErrNonUniqueUpdatesSubaccount
+		}
+
+		result, err := isValidIsolatedPerpetualUpdates(u, perpIdToMarketType)
+		if err != nil {
+			return false, nil, err
+		}
+		if result != types.Success {
+			success = false
+		}
+
+		successPerUpdate[i] = result
+		idOfSettledUpdates[*u.SettledSubaccount.Id] = struct{}{}
+	}
+
+	return success, successPerUpdate, nil
+}
+
+// Checks whether the perpetual updates to a settled subaccount violates constraints for isolated
+// perpetuals. This function assumes the settled subaccount is valid and does not violate the
+// the constraints.
+// The constraint being checked is:
+//   - a subaccount with a position in an isolated perpetual cannot have updates other perpetuals
+//   - a subaccount with a position in a non-isolated perpetual cannot have updates for isolated
+//     perpetuals
+//   - a subaccount with no positions cannot be updated to have positions in multiple isolated
+//     perpetuals or a combination of isolated and non-isolated perpetuals
+func isValidIsolatedPerpetualUpdates(
+	settledUpdate settledUpdate,
+	perpIdToMarketType map[uint32]perptypes.PerpetualMarketType,
+) (types.UpdateResult, error) {
+	// If there are no perpetual updates, then this update does not violate constraints for isolated
+	// markets.
+	if len(settledUpdate.PerpetualUpdates) == 0 {
+		return types.Success, nil
+	}
+
+	// Check if the updates contain an update to an isolated perpetual.
+	hasIsolatedUpdate := false
+	isolatedUpdatePerpetualId := uint32(0)
+	for _, perpetualUpdate := range settledUpdate.PerpetualUpdates {
+		marketType, exists := perpIdToMarketType[perpetualUpdate.PerpetualId]
+		if !exists {
+			return types.UpdateCausedError, errorsmod.Wrap(
+				perptypes.ErrPerpetualDoesNotExist, lib.UintToString(perpetualUpdate.PerpetualId),
+			)
+		}
+
+		if marketType == perptypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_ISOLATED {
+			hasIsolatedUpdate = true
+			isolatedUpdatePerpetualId = perpetualUpdate.PerpetualId
+		}
+	}
```

Comment:
Do we need to add a constraint that if we see multiple isolated perpetual updates, that would be erroneous?

Reply:
To have multiple perpetual update entires for the same isolated perpetual, the length of the slice of perpetual updates would have to be > 1, given that each [PerpetualUpdate](https://github.com/dydxprotocol/v4-chain/blob/340801a3cc8013a8c6b61a0a3717b48a0fe70aeb/protocol/x/subaccounts/types/update.go#L89-L95) holds only the update for a single perpetual id.